File: Mika/branches/precompiled/core-vm/java/java/io/File.java
/**************************************************************************
* Parts copyright (c) 2001 by Punch Telematix. All rights reserved.       *
* Parts copyright (c) 2003 by Chris Gray, /k/ Embedded Java Solutions.    *
* All rights reserved.                                                    *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of Punch Telematix or of /k/ Embedded Java Solutions*
*    nor the names of other contributors may be used to endorse or promote*
*    products derived from this software without specific prior written   *
*    permission.                                                          *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL PUNCH TELEMATIX, /K/ EMBEDDED JAVA SOLUTIONS OR OTHER *
* CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,   *
* EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,     *
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR      *
* PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  *
* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    *
* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS      *
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.            *
**************************************************************************/

package java.io;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.NoSuchElementException;
import java.util.Random;
import java.util.StringTokenizer;
import java.util.Vector;

public class File implements Comparable, Serializable {

  private static final long serialVersionUID = 301077366599181567L;

  static Vector toBeDeleted;
  
  public static final String separator;
  public static final char   separatorChar;
  public static final String pathSeparator;
  public static final char   pathSeparatorChar;

  private static String current_working_dir = null;
  private static String fsroot_dir = null;
  private static String fsrootedPrefix;

  /**
   ** Name by which the File object was created ('path' parameter of the constructor).
   */
  private transient  String     fullname;
  /*
  ** If the path begins with '/', this is stripped off and 'absolute' is set.
  ** If the path begins with '{}/', this is stripped off and 'fsrooted' is set.
  ** The remainder of the path is then split at the last '/':
  **  - everything up to the last '/' (exclusive) goes in 'dirpath'.
  **  - everything after the last '/' (exclusive) goes in 'filename'.
  */
  private transient  boolean    absolute;
  private transient  boolean    fsrooted;
  private transient  String     dirpath;
  private transient  String     filename;

  /**
   ** The absolute path used to open the file, after path mungeing.
   */
  private transient  String     absolutePath;
  /**
   ** Same as absolutePath, but after eliminating './' and '../' sequences.
   ** Generated on demand by toURL().
   */
  private transient String canonicalPath;

  /**
   ** Same as absolutePath. Used by the native functions to access the file.
   */
  final transient String absname;

  private static native String get_CWD();
  private static native String get_fsroot();
  private static native void init();
  
  static {
    init();
    
    separator = GetSystemProperty.FILE_SEPARATOR;
    pathSeparator = GetSystemProperty.PATH_SEPARATOR;
    separatorChar = separator.charAt(0);
    pathSeparatorChar = separator.charAt(0);
    fsrootedPrefix = "{}" + separator;

    current_working_dir = get_CWD();
    fsroot_dir = get_fsroot() + separator;
    current_working_dir = current_working_dir + separator;
    System.getProperties().setProperty("user.dir", current_working_dir);
  }

  public static File[] listRoots() {
    return new File[] { new File("/") };
  }

  public static File createTempFile(String pre, String suf) throws IOException {
    return createTempFile(pre, suf, null);
  }

  public static File createTempFile(String pre, String suf, File dir) throws IOException {
    if(dir == null) {
      dir = new File(System.getProperties().getProperty("java.io.tmpdir", "/tmp"));
    }
    
    if(suf == null) {
      suf = ".tmp";
    }

    if(pre == null) {
      pre = "";
    }

    File result = new File(dir, pre + (new Random()).nextLong() + suf);
    result.createNewFile();

    return result;
  }

  private String pack(String path) {
    String result = "";
    LinkedList list = new LinkedList();
    StringTokenizer st = new StringTokenizer(path, "/");

     try {
      while (true) {
        String token = st.nextToken();
        if(!token.equals(".")) { 
          if(!token.equals("..")) {
            list.add(token);
          } else {
            if(list.size() > 0)
              list.removeLast();
          }
        }
      }
    }
    catch(NoSuchElementException e) {
    }

    Iterator iter = list.iterator();
    while(iter.hasNext()) 
      result += "/" + (String)iter.next();
      
    if(!path.startsWith("/") && (result.length() > 0)) result = result.substring(1);
    //if(path.endsWith("/")) result += "/";

    return result; 
  }

  private String stripSlashes(String path) {
    String result = path;
    int i = result.indexOf("//");

    while (i >= 0) {
      result  = result.substring(0, i) + result.substring(i + 1);
      i = result.indexOf("//");
    }
    return result;
  }

  public File(String path) throws NullPointerException {
    String relpath;
    filename = "";

    if(path.length() == 0) path = current_working_dir;

    if(path.charAt(0) == separatorChar) {  // Absolute path
      relpath = path.substring(1);
      absolute = true;
    } else if(path.startsWith(fsrootedPrefix)) {  // fsrooted path
      relpath = path.substring(3);
      fsrooted = true;
    } else {
      relpath = path;
    }

    try {

      int dirpathlen = relpath.lastIndexOf(separatorChar);
      if (dirpathlen < 0) {
        dirpath = "";
        filename = relpath;
      } else {
        dirpath = relpath.substring(0, dirpathlen);
        filename = relpath.substring(dirpathlen + 1);
      }
    } catch (IndexOutOfBoundsException e) {
      e.printStackTrace();
    }

    fullname = stripSlashes(path);
    if (absolute) {
      absolutePath = fullname;
    }
    else if (fsrooted) {
      absolutePath = fsroot_dir + fullname.substring(3);
    }
    else {
      absolutePath = current_working_dir + fullname;
    }
    int len = absolutePath.length() - 1;
    if(len >= 0 && absolutePath.charAt(len) == '/') {
      absolutePath = absolutePath.substring(0, len);
    }

    
    absname = absolutePath;
  }

  public File(String dirname, String name) throws NullPointerException {
    this(name == null ? null : (dirname == null ? "" : (dirname.equals("") ? "" : dirname + separatorChar)) + name);
  }

  public File(File dir, String name) throws NullPointerException {
    this((dir == null ? "" : dir.getPath()), name);
  }

  public String toString() {
    return fullname;
  }

  public boolean equals(Object obj) {
    if ((obj != null) && obj instanceof File) {
      return (((File)obj).getPath().equals(getPath()));
    }
    return false;
  }

  public int compareTo(File pathname) {
    return (getPath().compareTo(pathname.getPath()));
  }

  public int compareTo(Object obj) {
    if (obj instanceof File) {
      return (absolutePath.compareTo(((File)obj).getAbsolutePath()));
    } else { throw new ClassCastException(); }
  }

  public int hashCode() {
    return getPath().hashCode() ^ 1234321;
  }

  public String getName() {
    String result = new String(filename);
    int snip = filename.lastIndexOf(separatorChar);
    if (snip >= 0) result = result.substring(snip+1);

    return result;
  }

  public String getPath() {
    return fullname;
  }

  public String getAbsolutePath() {
    return absolutePath;
  }

  public File getAbsoluteFile() {
    return new File(absolutePath);
  }

  public String getCanonicalPath() throws IOException {
    if(canonicalPath == null) {
      canonicalPath = pack(absolutePath);
    }
    return canonicalPath;
  }

  public File getCanonicalFile() throws IOException {
    return new File(getCanonicalPath());
  }

  public String getParent() {
    if (dirpath.length() + filename.length() == 0) {

      return null;

    }

    String prefix = absolute ? separator : fsrooted ? fsrootedPrefix : "";

    if (filename.length() == 0) {

      return new File(prefix + dirpath).getParent();
      
    }

    return prefix + dirpath;
  }

  public File getParentFile() {
    String parentName = getParent();
 
    if (parentName == null) {
      return null;
    }
 
    return new File(parentName);
  }

  public boolean isAbsolute() {
    return absolute;
  }

  public boolean isHidden() {
    return (filename.charAt(0) == '.');
  }
  
  public boolean createNewFile() throws IOException {
    if(!this.exists()) {
      try {
        FileOutputStream fo = new FileOutputStream(this);
        fo.close();
        return true;
      } catch(FileNotFoundException e) {
        throw new IOException("Unable to create " + this);
      }
    }
    return false;
  }

  public native boolean exists() throws SecurityException;

  public native String[] list() throws SecurityException;

  public String[] list(FilenameFilter filter) throws SecurityException {
    String[] files = this.list();
    if(files == null){
      return null;
    }

    ArrayList alist = new ArrayList(files.length);
    for(int i=0; i < files.length; i++){
      if(filter.accept(this, files[i])){
        alist.add(files[i]);
      }
    }
    String[] result = new String[alist.size()];
    return (String[])alist.toArray(result);
  }

  public File[] listFiles() throws SecurityException {
    String[] filenames = this.list();
    if(filenames == null){
      return null;
    }
    File[] files = new File[filenames.length];
    for(int i=0; i<filenames.length; i++) 
      files[i] = new File(this, filenames[i]);
    return files;
  }

  public File[] listFiles(FilenameFilter filter) throws SecurityException {
    String[] filenames = this.list(filter);
    if(filenames == null){
      return null;
    }
    File[] files = new File[filenames.length];
    for(int i=0; i<filenames.length; i++)
      files[i] = new File(this, filenames[i]);
    return files;
  }

  public File[] listFiles(FileFilter filter) throws SecurityException {
    File[] files = this.listFiles();
    if(files == null){
      return null;
    }
    ArrayList alist = new ArrayList(files.length);
    for(int i=0; i < files.length; i++){
      if(filter.accept(files[i])){
        alist.add(files[i]);
      }
    }
    File[] result = new File[alist.size()];
    return (File[])alist.toArray(result);
  }

  public native boolean canRead() throws SecurityException;

  public native boolean canWrite() throws SecurityException;

  public native boolean isFile() throws SecurityException;

  public native boolean isDirectory() throws SecurityException;

  public native long lastModified() throws SecurityException;

  public native boolean setLastModified(long time) throws SecurityException;

  public void deleteOnExit() throws SecurityException {
    if(toBeDeleted == null) {
      createShutdownHook();
    }
    toBeDeleted.add(this);
  }

  private synchronized static void  createShutdownHook(){
    if(toBeDeleted == null) {
      toBeDeleted = new Vector();
      Runtime.getRuntime().addShutdownHook(new Thread("File:deleteOnExit") {
        public void run() {
          Iterator iter = File.toBeDeleted.iterator();
          while(iter.hasNext()) {
            File f = (File)iter.next();
            try {
              if(f != null) f.delete();
            }
            catch(Exception e) {
            }
          }
        }
      });
    }
  }

  public boolean setReadOnly() throws SecurityException {
    return false;
  }

  public native long length() throws SecurityException;

  public native boolean mkdir() throws SecurityException;

  public boolean mkdirs() throws SecurityException {
    boolean result = false;
    if(!absolutePath.equals(separator) && !absolutePath.equals(fsrootedPrefix)) {
      File parent = getParentFile();
      if (parent != null) {
        parent.mkdirs();
      }
      result = mkdir();
    }
    return result;
  }

  private native boolean rename(String src, String dest);
  
  public boolean renameTo(File dest) throws SecurityException {
    return rename(absolutePath, dest.getAbsolutePath());
  }

  public native boolean delete() throws SecurityException;

  public URL toURL() throws MalformedURLException {
    if(canonicalPath == null) {
      canonicalPath = pack(absolutePath);
    }
    return new URL("file", "", canonicalPath);
  }

}

File: core-vm/src/native/java/io/File.c
/**************************************************************************
* Copyright (c) 2021, 2022, 2023 by KIFFER Ltd. All rights reserved.      *
*                                                                         *
* Redistribution and use in source and binary forms, with or without      *
* modification, are permitted provided that the following conditions      *
* are met:                                                                *
* 1. Redistributions of source code must retain the above copyright       *
*    notice, this list of conditions and the following disclaimer.        *
* 2. Redistributions in binary form must reproduce the above copyright    *
*    notice, this list of conditions and the following disclaimer in the  *
*    documentation and/or other materials provided with the distribution. *
* 3. Neither the name of KIFFER Ltd nor the names of other contributors   *
*    may be used to endorse or promote products derived from this         *
*    software without specific prior written permission.                  *
*                                                                         *
* THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED          *
* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF    *
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.    *
* IN NO EVENT SHALL KIFFER LTD OR OTHER CONTRIBUTORS BE LIABLE FOR ANY    *
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL      *
* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE       *
* GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS           *
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER    *
* IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR         *
* OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF  *
* ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                              *
**************************************************************************/

#ifdef FREERTOS
#include <ff_stdio.h>
#endif

#include "vfs.h"
#include "core-classes.h"
#include "exception.h"
#include "fields.h"
#include "wordset.h"
#include "wstrings.h"

/*
** Get the file's name as a C-style UTF8 string. After you've finished with the
** name, call releaseMem() on the result to free up the memory.
*/
char *getFileName(w_instance thisFile) {
  w_instance pathString;
  w_string path;

  pathString = getReferenceField(thisFile, F_File_hostpath);
  path = String2string(pathString);

  return w_string2UTF8(path, NULL);	  
}

w_boolean statFile(w_instance thisFile, struct vfs_stat_t *statbufptr) {
  char *pathname;
  w_boolean result;
  
  pathname = getFileName(thisFile);	  

  result = (vfs_stat(pathname, statbufptr) != -1);

  releaseMem(pathname);

  return result;
}

w_boolean File_createNew (w_thread thread, w_instance thisFile) {
  struct vfs_stat_t statbuf;
  struct vfs_FILE *file;
  char *pathname;
  w_boolean result;
  
  pathname = getFileName(thisFile);	  
  result = !!vfs_stat(pathname, &statbuf);
  if (result) {
    file = vfs_open(pathname, VFS_O_CREAT, VFS_S_IRWXU | VFS_S_IRGRP | VFS_S_IROTH);
    if (file) {
      vfs_close(file);
    }
    else {
      throwException(thread, clazzIOException, "could not open file '%s' using mode w+: %s\n", pathname, strerror(errno));
    }
  }

  releaseMem(pathname);

  return result;
}

w_boolean File_exists (w_thread thread, w_instance thisFile) {
  struct vfs_stat_t statbuf;

  return statFile(thisFile, &statbuf);
}

/**
 * Get the current working directory (without any trailing slash).
 * When running under an RTOS this will return NULL, and the Java code will use the value of
 * the user.dir property instead.
 */
w_instance File_get_CWD (w_thread thread, w_instance thisFile) {
#ifdef FREERTOS
  return NULL;
#else
  w_int cwdlen = strlen(current_working_dir);
  if (current_working_dir[cwdlen-1] == '/') {
    --cwdlen;
  }
  return getStringInstance(utf2String(current_working_dir, cwdlen));
#endif
}

w_instance File_get_fsroot (w_thread thread, w_instance thisFile) {
  // special case: fsroot = "/" really means fsroot = "", i.e. "{}/" maps to "/" (not "//")
  w_string fsroot_string = (strlen(fsroot) == 1 && fsroot[0] == '/') ? registerString(string_empty) : utf2String(fsroot, strlen(fsroot));
  return getStringInstance(fsroot_string);
}

#ifdef FREERTOS
// FreeRTOS FAT filesystem uses a different logic for scanning directories:
// ff_findfirst is equivalent to opendir followed by readdir
// TODO can we rewrite the POSIX code to work the same way?

w_instance File_list (w_thread thread, w_instance thisFile) {
  char *pathname;
  int count=0;
  int i;
  w_instance result = NULL;
  char *entryname;
  w_string entry_string;
  w_instance entry;
  w_wordset ws = NULL;
  w_wordset *temp = &ws;
  w_size numberOfFiles = 0;
  struct vfs_stat_t statbuf;
  FF_FindData_t finddata_buffer;

  threadMustBeSafe(thread);
  pathname = getFileName(thisFile);	  
  woempa(1, "dir %s\n", pathname);

  if (ff_stat(pathname, &statbuf) != 0 || statbuf.st_mode != FF_IFDIR) {
    // this is not a directory
    releaseMem(pathname);
    return NULL;
  }

  if (ff_findfirst(pathname, &finddata_buffer ) == 0 ) {
    do {
      const int fileNameLength = strlen(finddata_buffer.pcFileName);
      if (fileNameLength < 1 || finddata_buffer.pcFileName[0] !='.' || finddata_buffer.pcFileName[fileNameLength-1] != 0 || fileNameLength > 2) {
        entryname = allocMem(fileNameLength + 1);
        strcpy(entryname, finddata_buffer.pcFileName);
        woempa(1, "  put entry %s\n", entryname);
      }
      else {
        woempa(1, "Dir entry '%s' is bogus, ignoring\n", finddata_buffer.pcFileName);
      }
      addToWordset(temp, (w_word) entryname);
    } while( ff_findnext(&finddata_buffer) == 0 );
  }
  releaseMem(pathname);
    
  if (!wordsetIsEmpty(temp)) {
    enterUnsafeRegion(thread);
    result = allocArrayInstance_1d(thread, clazzArrayOf_String, sizeOfWordset(temp));
    enterSafeRegion(thread);
  }

  for (i = 0; !wordsetIsEmpty(temp); ++i) {
    entryname = (char*) takeFirstFromWordset(temp);
    woempa(1, "  take entry %s\n", entryname);
    entry_string = utf2String(entryname, strlen(entryname));
    releaseMem(entryname);
    entry = getStringInstance(entry_string);
    setArrayReferenceField(result, entry, i);
    deregisterString(entry_string);
  }
  releaseWordset(temp);

  return result;
}

#else

// POSIX version

w_instance File_list (w_thread thread, w_instance thisFile) {
  char *pathname;
  int count=0;
  int i;
  w_instance result = NULL;
  char *entryname;
  w_string entry_string;
  w_instance entry;
  vfs_dir_t       *dir;
  vfs_dirent      *dirent;
  
  threadMustBeSafe(thread);
  pathname = getFileName(thisFile);	  

  dir = vfs_opendir(pathname);
  
  releaseMem(pathname);

  if(dir != NULL) {

    /* Count the entries */
    while((dirent = vfs_readdir(dir)) != NULL) count++;
    
    /* Substract 2 for . and .. */
    count -= 2;

    if(count < 0) {
      count = 0;
      woempa(9, "!!! %s is a corrupt directory, attempting to read it anyway...\n", pathname);
      woempa(9, "!!! It's recommended do to a fsck -f on this partition as soon as possible to prevent further corruption\n");
    }
    
    enterUnsafeRegion(thread);
    result = allocArrayInstance_1d(thread, clazzArrayOf_String, count);
    enterSafeRegion(thread);

    /* Rewind the directory */
    vfs_rewinddir(dir);
    i = 0;
    
    /* Fill up the array - don't trust vsf_readdir not to return more elements */
    while((dirent = vfs_readdir(dir)) != NULL && i < count) {
      entryname = dirent->d_name;
      if (strcmp(entryname, ".") && strcmp(entryname, "..")) {
        entry_string = utf2String(entryname, strlen(entryname));
        entry = getStringInstance(entry_string);
        setArrayReferenceField(result, entry, i++);
        deregisterString(entry_string);
        removeLocalReference(thread, entry);
      }	
    }
  
    vfs_closedir(dir);
  }

  return result;

}
#endif

w_boolean File_setReadOnly (w_thread thread, w_instance thisFile) {
  char *pathname;
  struct vfs_stat_t statbuf;
  w_boolean result;
  const int WRITABLE = VFS_S_IWUSR | VFS_S_IWGRP | VFS_S_IWOTH;
  
  pathname = getFileName(thisFile);	  
  result = vfs_stat(pathname, &statbuf) == 0
    && vfs_chmod(pathname, statbuf.st_mode & ~WRITABLE) == 0;

  releaseMem(pathname);
  
  return result;
}

w_boolean File_canRead (w_thread thread, w_instance thisFile) {
  struct vfs_stat_t statbuf;
  
  return statFile(thisFile, &statbuf) && (((statbuf.st_mode & VFS_S_IRWXU) & VFS_S_IRUSR) == VFS_S_IRUSR);
}

w_boolean File_canWrite (w_thread thread, w_instance thisFile) {
  struct vfs_stat_t statbuf;
  
  return statFile(thisFile, &statbuf) && (((statbuf.st_mode & VFS_S_IRWXU) & VFS_S_IWUSR) == VFS_S_IWUSR);
}

w_boolean File_isFile (w_thread thread, w_instance thisFile) {
  struct vfs_stat_t statbuf;
  
  return statFile(thisFile, &statbuf) && VFS_S_ISREG(statbuf.st_mode);
}

w_boolean File_isDirectory (w_thread thread, jobject thisFile) {
  struct vfs_stat_t statbuf;
  
  return statFile(thisFile, &statbuf) && VFS_S_ISDIR(statbuf.st_mode);
}

w_long File_lastModified (w_thread thread, jobject thisFile) {
  struct vfs_stat_t statbuf;

/*
TODO sort out why  ffconfigTIME_SUPPORT == 1 and yet no st_mtime in statbuf
#if ( ffconfigTIME_SUPPORT == 1 )
  if (statFile(thisFile, &statbuf)) {
    return ((jlong)statbuf.st_mtime) * ((jlong)1000);
  }
*/
    return 0;
}

w_long File_length (w_thread thread, jobject thisFile) {
  struct vfs_stat_t statbuf;
  
  if (statFile(thisFile, &statbuf)) {
    return statbuf.st_size;
  }
  else {
    return 0;
  }
}

w_boolean File_delete (w_thread thread, w_instance thisFile) {
  char      *pathname;
  struct vfs_stat_t statbuf;
  w_boolean result;
  
  pathname = getFileName(thisFile);	  

  result = (vfs_stat(pathname, &statbuf) == 0);
  if (result) {
    if(VFS_S_ISDIR(statbuf.st_mode)) {
      result = (vfs_rmdir(pathname) == 0);
      woempa(9, "%s is a directory, result = %d\n", pathname, result);
    } else {
      result = (vfs_unlink(pathname) == 0);
      woempa(9, "%s is a file, result = %d\n", pathname, result);
    }
  }
  
  releaseMem(pathname);

  return result;

}

w_boolean File_mkdir(w_thread thread, jobject thisFile) {
  char *pathname;
  int rc;
  w_boolean result;
  
  pathname = getFileName(thisFile);	  

  rc = vfs_mkdir(pathname, VFS_S_IRWXU | VFS_S_IRWXG | VFS_S_IRWXO);
  result = (rc != -1);

  releaseMem(pathname);

  return result;
}

w_boolean File_setModTime(w_thread thread, w_instance thisFile, w_long modtime) {
  char *pathname;
  w_boolean        result;
  struct utimbuf  buf;

  pathname = getFileName(thisFile);
  modtime = modtime / 1000;

  buf.actime  = (time_t) modtime; 
  buf.modtime = (time_t) modtime; 

  result = (0 == vfs_utime(pathname, &buf));

  releaseMem(pathname);

  return result;
}

w_boolean File_rename(w_thread thread, w_instance thisObj, w_instance file1String, w_instance file2String) {
    
  w_string file1_string;
  w_string file2_string;
  char *pathname1;
  char *pathname2;
  w_int pathlength1;
  w_int pathlength2;
  w_boolean        result = FALSE;

  if (!file1String || !file2String) {
    throwException(thread, clazzNullPointerException, NULL);
  }

  file1_string = String2string(file1String);
  file2_string = String2string(file2String);
  pathname1 = w_string2UTF8(file1_string, &pathlength1);
  pathname2 = w_string2UTF8(file2_string, &pathlength2);

  result = (vfs_rename(pathname1, pathname2) == 0);

  releaseMem(pathname1);
  releaseMem(pathname2);

  return result;

}

